### ISSUE TITLE
Update Doxygen version used on docs.wxwidgets.org

### ISSUE THREAD

**Issue by wxtrac** [wxWidgets/wxWidgets#19126]
*Issue migrated from trac ticket # 19126*

**component:** documentation | **priority:** normal

#### 2021-03-31 02:59:13: @vadz created the issue

___
The current version has many bugs, [here is the latest one](https://github.com/wxWidgets/wxWidgets/commit/da8c851949a0b9f6658379c264094b6909802bbf#commitcomment-48836833) with a spurious line break after `(e.g.`.

We should upgrade it to something closer to the current version, even if not necessarily the latest one.

**Comment by EwoutH** [comment]
@vadz @tierra @dconnet Thanks for creating and working on this issue! Has any progress been made so far upgrading the Doxygen version?

**Comment by vadz** [comment]
No, sorry, I don't think anybody has looked into this.

I wonder if we could use GitHub Pages for the generated docs -- and Actions to update them. If anybody would consider looking into this, it would be really great.

**Comment by tierra** [comment]
I outlined the brief investigation I did with GitHub Pages for docs back in Sept:
https://groups.google.com/g/wx-dev/c/CwrIAAuEMCU/m/3T20wsufCgAJ

tl;dr: It's possible, but with a few concessions that we never really discussed there.

That said, the idea of using Actions to update the trunk docs, where it's currently hosted rather than moving it to GitHub Pages, is still a pretty good idea itself. It would decouple the version of Doxygen from what can be built on the server hosting the docs (which as I mentioned already, stops working after version 1.8.15 currently). It can be setup pretty much exactly the same way the official site is, with a secret key configured in the repo.

Like the main website, building that job on actions really boils the server(s) hosting those sites down to pretty much anything running Apache (for redirects and error page configs), PHP (for doc search, and a few other things), adequate disk space, and the ability to sync files over SSH. It also of course empowers anyone to reconfigure and update it easily, even without access to the server.

**Comment by tierra** [comment]
> I outlined the brief investigation I did with GitHub Pages for docs back in Sept: https://groups.google.com/g/wx-dev/c/CwrIAAuEMCU/m/3T20wsufCgAJ
> 
> tl;dr: It's possible, but with a few concessions that we never really discussed there.

Since this was in the next email, and might be missed: The biggest of these concessions is that GitHub Pages is restricted to 1GB max, and the current collection of manuals on the site right now come out to 3GB. So it would require removing nearly 70% of the old versions hosted. This is pretty bad though since the manual has only grown over time, and even the latest is 216MB alone, so eventually it'd strangle us down to only 4 different versions max, likely only hosting the latest per major version branch.

```
$ du -smch *
0       2.4
14M     2.4.2
0       2.6
6.4M    2.6.3
6.4M    2.6.4
0       2.8
7.7M    2.8.0
13M     2.8.12
7.8M    2.8.4
13M     2.8.8
0       2.9
79M     2.9.1
104M    2.9.3
163M    2.9.5
0       3.0
163M    3.0.0
164M    3.0.1
164M    3.0.2
192M    3.0.3
193M    3.0.4
201M    3.0.5
0       3.1
187M    3.1.0
203M    3.1.1
205M    3.1.2
207M    3.1.3
210M    3.1.4
213M    3.1.5
4.0K    404.php
26M     docsets
4.0K    favicon.ico
8.0K    index.html
4.0K    robots.txt
0       stable
4.0K    style.css
216M    trunk
2.9G    total
```

At the least, you'd obviously want 2.8.12, 3.0.5, 3.1.5, and trunk, and that's 630MB right there.

Anyway, like I said above, this isn't really critical though. Docs can be hosted pretty much anywhere with the bare minimum of Apache, PHP, ~5GB of space, and SSH. That's painless to find and setup (and just fine where it is right now).

**Comment by tierra** [comment]
FYI: If anyone does work on this, and builds a new Actions workflow (with most of the same upload logic as the @wxWidgets/website repo), I'd be happy to configure the SSH key and tear down the existing cron job on the server, since those tasks do require access.

**Comment by vadz** [comment]
@tierra I've setup a new key on the server which is stored in `DOCS_WEBSITE_KEY` secret on GitHub. I've also added `DOCS_WEBSITE_SSH_CONFIG` secret which contains the `.ssh/config` stanza sufficient for `rsync ... wxdocs:` to work and use it in [my new workflow](https://github.com/vadz/wxWidgets/blob/docs-workflow/.github/workflows/docs_update.yml).

Testing this (in my own repository for now) it seems to work fine, see https://docs.wxwidgets.org/latest/

However search still doesn't work and I'm not sure what needs to be done to make it work with Doxygen 1.8.17 used for these docs -- I hoped it would just work on its own with the newer version.

Also, we've lost the analytics script in the footer, any idea how to handle this? I don't think it's secret (considering that it is part of the publicly visible HTML), so should I put it in the workflow file itself or do you see something better?

**Comment by vadz** [comment]
Sorry, search actually does seem to work, see e.g. https://docs.wxwidgets.org/latest/search.php?query=wildcard for a simple example.

But the search results are below the screen and not visible by default, which is why I thought it didn't work initially. This is clearly due to something that we do in `docs/doxygen/extra_stylesheet.css`, e.g. removing these lines https://github.com/wxWidgets/wxWidgets/blob/9dc7248b1d745ca1a47c4006081311f4553114f4/docs/doxygen/extra_stylesheet.css#L13-L15 solves the problem to me. But I'm not sure if it's the right thing to do and the search terms still appear in a weird place, so if someone more familiar with CSS than me could please have a look at it, it would be really great. TIA!

**Comment by vadz** [comment]
For now I'm merging the changes sufficient to use 1.8.17 which I copied to docs.wxwidgets.org (@tierra If you prefer, please remove the binary I copied there and just run `apt install doxygen`, as it's exactly the version available in Ubuntu 20.04 LTS which is used on this machine currently).

We have #18849 for the CSS issues and I'll add a GitHub workflow too, but for now let's switch to 1.8.17 already.

**Comment by vadz** [comment]
Damn, and of course this broke everything with the header not looking good at all any more :-( I have no idea why I didn't notice this while testing locally.

**Comment by vadz** [comment]
I have no idea what is going on here... I've tried using a different Doxygen version (1.9.3), but I still don't see the problem locally, where everything looks fine, while it's still broken on docs.wxwidgets.org. There is also a "Toggle main menu visibility" checkbox, coming from Doxygen `menu.js`, visible online, which I don't see at all when testing locally.

I've already purged Cloudflare cache twice, so this must not be it, but what else can it be?

**Comment by vadz** [comment]
I've regenerated the docs using 1.8.8 now and stopped the cron job on the server as it's not going to work with the latest master and I don't want to rollback the recent changes yet, as I still hope to make them work with the new Doxygen. But right now I'm completely stumped and just don't understand why things work locally for me but not on the server, using exactly the same Doxygen binary.

**Comment by vadz** [comment]
I feel pretty sheepish, but after double- and triple-checking Cloudflare, it looks like the problem was actually with my local browser cache because I don't see it any longer after clearing it.

Please let me know if anybody else sees the problem, but for me everything finally seems to work fine with 1.8.17 on the server (except for the CSS problem mentioned above).

**Comment by tierra** [comment]
Sorry I missed out on the fun, but it looks like you got everything working great.

This is all still using the Doxygen binary on the server, using the cronjob, right? Haven't switched to the new workflow yet?

**Comment by tierra** [comment]
The styles look like they're working fine for search too.

**Comment by tierra** [comment]
As for the analytics, I just always kept that around in a local git commit on the server there, as a change to the `custom_footer.html` file. The update script would just `git pull --rebase` to keep that applied before regenerating. It looks like you might have changed that though?

```
# Undo any changes previously done to the custom footer.
git checkout -- docs/doxygen/custom_footer.html
```

**Comment by tierra** [comment]
Oh, I see, you just inject it with `sed` now. That works.

**Comment by vadz** [comment]
Yes, I've got rid of this custom commit as it broke docs rebuild twice while I was making my changes (once when it got in conflict with the new placeholder comment I added and another time when I've added a copy command using `out/html` -- which, of course, was wrong on the server), so now it should be more robust. But I think we should still to GitHub Actions way of updating the docs because it's both faster and more efficient (we update the docs more rarely than daily, but it's nice to have to wait for up to 24 hours when we do update them) and will just switch `latest` to `trunk` and stop the cron job soon if it works without problems.

This will also make it easier to change the docs generation in the future as the workflow file is more visible/accessible than the cron script on the server. In particular, I'm still thinking about switching to the "awesome" (I hate the name, but like almost everything else) Doxygen theme, and this would require using Doxygen 1.9.x for best results.

**Comment by tierra** [comment]
Yeah, having that trigger to rebuild on push is helpful.

I like that theme better, but I also know how painful doing these Doxygen upgrades are, primarily because there really doesn't seem to be much of a backwards compatibility effort/policy, even with minor version Doxygen updates. I'm sure that 3rd party theme is even more likely to break on upgrades than just our custom styles/header/footer.

If you do check out that theme, keep an eye on our custom aliases, such as the event/flag tables and screenshots (appearance), etc. The theme might not handle those as cleanly out of the box as the rest of the standard Doxygen markup.
